packages/SwingSet/src/controller/controller.js

```diff
@@ -446,15 +446,35 @@ export async function makeSwingsetController(
     /**
      * @param {string} vatName
      * @param {string} method
-     * @param {import('@endo/marshal').CapData<unknown>} args
+     * @param {unknown[]} args
      * @param {ResolutionPolicy} resultPolicy
+     * @param {string[]} [slots]
      */
-    queueToVatRoot(vatName, method, args, resultPolicy = 'ignore') {
+    queueToVatRoot(
+      vatName,
+      method,
+      args = [],
+      resultPolicy = 'ignore',
+      slots = [],
+    ) {
+      function replacer(_, arg) {
+        if (typeof arg === 'bigint') {
+          return { [QCLASS]: 'bigint', digits: String(arg) };
+        }
+        if (arg === undefined) {
+          return { [QCLASS]: 'undefined' };
+        }
+        return arg;
+      }
+
       const vatID = kernel.vatNameToID(vatName);
       assert.typeof(method, 'string');
-      insistCapData(args);
       const kref = kernel.getRootObject(vatID);
-      const kpid = kernel.queueToKref(kref, method, args, resultPolicy);
+      const methargs = {
+        body: JSON.stringify([method, args], replacer),
```

Comment:
Previously callers of `queueToVatRoot` had to provide a pre-cooked capdata object (and of course creating a capdata requires a `JSON.stringify` call to produce the `body` string).  `queueToVatRoot` is used by tests to inject messages into the kernel from the outside.  This particular PR required updating a lot of tests, so one of the things I did was move some of the necessary quasi-serialization work that previously each test had to do by hand into `queueToVatRoot` in order to make the API less delicate.  Ideally we would just use marshal and avoid a lot of the mess here, but unfortunately, we can't.  This entry point has to weirdly straddle the world as seen from inside a vat and world as seen from inside the kernel -- it's used to send messages into a vat as if the caller was another vat sending the message, but the caller isn't in a vat; it's a piece of unit test code that in effect exists in kernel space.  In particular, the stuff that would otherwise be done by `convertValToSlot` must instead be done by providing the slots array (containing raw kref strings) directly, because there is no clist for the sender.  Tests can get away with this because they construct their worlds to be just so and the worlds they construct tend to be very simple.

The replacer you asked about is migrated from some of the tests which sent messages with arguments containing (in this case) bigints and undefineds. and this freed the callers from having to construct `@qclass` structures manually.  This is another piece of making the API less delicate.  Conceivably other `@qclass` things might need to be added in the future if tests require them, but I don't expect that.